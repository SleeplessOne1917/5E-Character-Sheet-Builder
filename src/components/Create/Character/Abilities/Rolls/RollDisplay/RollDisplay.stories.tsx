import { ComponentMeta, ComponentStory } from '@storybook/react';
import RollDisplay, { RollDisplayProps } from './RollDisplay';

import { mockAbilities } from '../../MockAbilitiesStore';

export default {
	title: 'Components/Create/Character/Abilities/Rolls/RollDisplay',
	component: RollDisplay,
	args: {
		abilities: mockAbilities
	},
	argTypes: {
		roll: { type: 'function' },
		onSelectAbility: { type: 'function' }
	}
} as ComponentMeta<typeof RollDisplay>;

const Template: ComponentStory<typeof RollDisplay> = (
	args: RollDisplayProps
) => <RollDisplay {...args} />;

export const ButtonShowing = Template.bind({});

export const RollNoAbilitySelected = Template.bind({});
RollNoAbilitySelected.args = {
	rolls: [6, 4, 1, 1],
	total: 11
};

export const RollWithAbilitySelected = Template.bind({});
RollWithAbilitySelected.args = {
	rolls: [6, 6, 6, 3],
	total: 18,
	ability: 'con'
};
