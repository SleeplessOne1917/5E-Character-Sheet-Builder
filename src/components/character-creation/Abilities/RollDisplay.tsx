import { ChangeEventHandler, MouseEventHandler, memo } from 'react';

import { AbilityItem } from '../../../types/srd';
import SmallButton from '../../Button/SmallButton';
import classes from './RollDisplay.module.css';

type RollDisplayProps = {
	rolls?: number[] | null;
	abilities: AbilityItem[];
	roll: MouseEventHandler<HTMLButtonElement>;
	ability: string;
	total?: number;
	onSelectAbility: ChangeEventHandler<HTMLSelectElement>;
};

const RollDisplay = ({
	rolls,
	abilities,
	roll,
	onSelectAbility,
	ability,
	total
}: RollDisplayProps) => {
	if (rolls) {
		return (
			<div className={classes['display-container']}>
				<h1>{total}</h1>
				<div className={classes.rolls}>
					{rolls.map((roll, index) => (
						<div key={index} className={classes.roll}>
							{roll}
						</div>
					))}
				</div>
				<select onChange={onSelectAbility} value={ability}>
					<option value="blank">&mdash;</option>
					{abilities.map(a => (
						<option value={a.index} key={a.index}>
							{a.index.toUpperCase()}
						</option>
					))}
				</select>
			</div>
		);
	} else {
		return (
			<div className={classes['display-container']}>
				<h2>&mdash;</h2>
				<SmallButton
					style={{ padding: '1rem', fontSize: '1rem' }}
					positive
					onClick={roll}
				>
					Roll
				</SmallButton>
			</div>
		);
	}
};

export default memo(RollDisplay);
