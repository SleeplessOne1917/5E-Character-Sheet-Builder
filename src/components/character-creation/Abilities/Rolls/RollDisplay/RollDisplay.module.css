.display-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 30%;
}

.rolls {
    display: flex;
    justify-content: center;
    width: 100%;
    align-items: center;
}

.roll {
    width: 20px;
    height: 20px;
    background-color: #1c9aef;
    color: var(--heaven-light);
    display: flex;
    justify-content: center;
    align-items: center;
    margin: 0 0.2rem;
    border-radius: 0.25rem;
}

.roll:last-child {
    background-color: #999;
}

.display-container button {
    width: 100%;
}

.display-container select {
    margin-top: 0.5rem;
    width: 100%;
    font-size: 1.5rem;
    font-family: inherit;
    border: 1px solid var(--davy);
    background-color: var(--heaven-light);
}

.display-container select:focus-visible {
    outline: var(--focus-outline);
}

.total {
    font-size: 2.2rem;
    font-weight: bold;
    margin: 1rem 0;
}

@media screen and (min-width: 768px) {
    .display-container {
        width: 15%;
    }
}
