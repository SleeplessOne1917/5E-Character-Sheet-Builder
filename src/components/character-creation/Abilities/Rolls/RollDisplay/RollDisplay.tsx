import { MouseEventHandler, memo } from 'react';

import { AbilityItem } from '../../../../../types/srd';
import Button from '../../../../Button/Button';
import Select from '../../../../Select/Select';
import classes from './RollDisplay.module.css';

export type RollDisplayProps = {
	rolls?: number[] | null;
	abilities: AbilityItem[];
	roll: MouseEventHandler<HTMLButtonElement>;
	ability?: string | null;
	total?: number;
	selectTestId?: string;
	onSelectAbility: (value: string) => void;
};

const RollDisplay = ({
	rolls,
	abilities,
	roll,
	onSelectAbility,
	ability,
	total,
	selectTestId
}: RollDisplayProps): JSX.Element => {
	if (rolls) {
		return (
			<div className={classes['display-container']} data-testid="roll-display">
				<div className={classes.total} data-testid="roll-total">
					{total}
				</div>
				<div className={classes.rolls}>
					{rolls.map((roll, index) => (
						<div key={index} className={classes.roll} data-testid="roll-die">
							{roll}
						</div>
					))}
				</div>
				<Select
					fontSize="1.5rem"
					onChange={onSelectAbility as (value: string | number) => void}
					value={ability ? ability : 'blank'}
					options={[{ value: 'blank', label: '\u2014' }].concat(
						abilities.map(a => ({
							value: a.index,
							label: a.index.toUpperCase()
						}))
					)}
					testId={selectTestId}
				/>
			</div>
		);
	} else {
		return (
			<div className={classes['display-container']} data-testid="roll-display">
				<div className={classes.total}>&mdash;</div>
				<Button positive onClick={roll} size="medium">
					Roll
				</Button>
			</div>
		);
	}
};

export default memo(RollDisplay);
