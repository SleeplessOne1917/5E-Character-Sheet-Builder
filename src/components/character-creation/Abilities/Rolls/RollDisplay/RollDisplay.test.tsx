import * as stories from './RollDisplay.stories';

import { render, screen } from '@testing-library/react';

import { composeStories } from '@storybook/testing-react';

const { ButtonShowing, RollNoAbilitySelected, RollWithAbilitySelected } =
	composeStories(stories);

describe('renders correctly', () => {
	it('when button is showing', () => {
		render(<ButtonShowing />);

		expect(screen.getByTestId(/roll-display/i)).toMatchSnapshot();
	});

	it('when roll is showing', () => {
		render(<RollNoAbilitySelected />);

		expect(screen.getByTestId(/roll-display/i)).toMatchSnapshot();
	});

	it('when roll is showing and ability is selected', () => {
		render(<RollWithAbilitySelected />);

		expect(screen.getByTestId(/roll-display/i)).toMatchSnapshot();
	});
});
