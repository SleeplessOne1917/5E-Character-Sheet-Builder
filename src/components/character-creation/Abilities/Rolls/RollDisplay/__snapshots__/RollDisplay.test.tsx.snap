// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when button is showing 1`] = `
<div
  class="display-container"
  data-testid="roll-display"
>
  <div
    class="total"
  >
    —
  </div>
  <button
    class="button positive"
    style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
    type="button"
  >
    Roll
  </button>
</div>
`;

exports[`renders correctly when roll is showing 1`] = `
<div
  class="display-container"
  data-testid="roll-display"
>
  <div
    class="total"
    data-testid="roll-total"
  >
    11
  </div>
  <div
    class="rolls"
  >
    <div
      class="roll"
      data-testid="roll-die"
    >
      6
    </div>
    <div
      class="roll"
      data-testid="roll-die"
    >
      4
    </div>
    <div
      class="roll"
      data-testid="roll-die"
    >
      1
    </div>
    <div
      class="roll"
      data-testid="roll-die"
    >
      1
    </div>
  </div>
  <div
    class="container"
    data-testid="select"
    style="font-size: 1.5rem;"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="button"
      data-testid="select-button"
      type="button"
    >
      —
      <svg
        aria-hidden="true"
        class="icon"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <ul
      aria-activedescendant="—"
      class="list"
      data-testid="select-list"
      role="listbox"
      style="display: none;"
      tabindex="-1"
    >
      <li
        aria-selected="true"
        class="item selected"
        data-testid="data-blank"
        role="option"
        tabindex="0"
      >
        —
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-str"
        role="option"
        tabindex="0"
      >
        STR
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-dex"
        role="option"
        tabindex="0"
      >
        DEX
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-con"
        role="option"
        tabindex="0"
      >
        CON
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-int"
        role="option"
        tabindex="0"
      >
        INT
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-wis"
        role="option"
        tabindex="0"
      >
        WIS
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-cha"
        role="option"
        tabindex="0"
      >
        CHA
      </li>
    </ul>
  </div>
</div>
`;

exports[`renders correctly when roll is showing and ability is selected 1`] = `
<div
  class="display-container"
  data-testid="roll-display"
>
  <div
    class="total"
    data-testid="roll-total"
  >
    18
  </div>
  <div
    class="rolls"
  >
    <div
      class="roll"
      data-testid="roll-die"
    >
      6
    </div>
    <div
      class="roll"
      data-testid="roll-die"
    >
      6
    </div>
    <div
      class="roll"
      data-testid="roll-die"
    >
      6
    </div>
    <div
      class="roll"
      data-testid="roll-die"
    >
      3
    </div>
  </div>
  <div
    class="container"
    data-testid="select"
    style="font-size: 1.5rem;"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="button"
      data-testid="select-button"
      type="button"
    >
      CON
      <svg
        aria-hidden="true"
        class="icon"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <ul
      aria-activedescendant="CON"
      class="list"
      data-testid="select-list"
      role="listbox"
      style="display: none;"
      tabindex="-1"
    >
      <li
        aria-selected="false"
        class="item"
        data-testid="data-blank"
        role="option"
        tabindex="0"
      >
        —
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-str"
        role="option"
        tabindex="0"
      >
        STR
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-dex"
        role="option"
        tabindex="0"
      >
        DEX
      </li>
      <li
        aria-selected="true"
        class="item selected"
        data-testid="data-con"
        role="option"
        tabindex="0"
      >
        CON
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-int"
        role="option"
        tabindex="0"
      >
        INT
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-wis"
        role="option"
        tabindex="0"
      >
        WIS
      </li>
      <li
        aria-selected="false"
        class="item"
        data-testid="data-cha"
        role="option"
        tabindex="0"
      >
        CHA
      </li>
    </ul>
  </div>
</div>
`;
